[devtools-blog] Simulate CVD in the Blink Renderer (#9115)
google/WebFundamentals: src/content/en/updates/2020/11/cvd.md

project_path: /web/_project.yaml
book_path: /web/updates/_book.yaml
description: Simulating color vision deficiencies in the Blink Renderer. The Chrome DevTools engineering blog - by the developers who build the DevTools.

{# wf_updated_on: 2020-11-13 #}
{# wf_published_on: 2020-11-13 #}
{# wf_tags: devtools-blog #}
{# wf_featured_image: /web/updates/images/2020/11/cvd/cvd.png #}
{# wf_featured_snippet: Simulating color vision deficiencies in the Blink Renderer. #}
{# wf_blink_components: N/A #}

# Simulating color vision deficiencies in the Blink Renderer {: .page-title }

{% include "web/_shared/contributors/mathiasbynens.html" %}

This article describes why and how we implemented color vision deficiency simulation in DevTools and the Blink Renderer.

## Background: bad color contrast {: #background }

[Low-contrast text](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) is the most common automatically-detectable accessibility issue on the web.

> ![CDP](/web/updates/images/2020/11/cvd/common-a11y-issues.svg)

Reviewer: ```suggestion
![A list of common accessibility issues on the web. Low-contrast text is by far the most common issue.](/web/updates/images/2020/11/cvd/common-a11y-issues.svg)
```

<div class="video-wrapper-full-width">
  <iframe class="devsite-embedded-youtube-video" data-video-id="mK_XmFb8E_w"
          data-autohide="1" data-showinfo="0" frameborder="0" allowfullscreen>
  </iframe>
</div>

In Puppeteer, [the new `page.emulateVisionDeficiency(type)` API](https://github.com/puppeteer/puppeteer/blob/v5.4.1/docs/api.md#pageemulatevisiondeficiencytype) lets you programmatically enable these simulations.

## Color vision deficiencies {: #cvd }

[Roughly 1 in 20 people](http://www.colourblindawareness.org/colour-blindness/) suffer from a color vision deficiency (also known as the less accurate term “color blindness”). Such impairments make it harder to tell different colors apart, which *can amplify contrast issues*.

<figure>
    <picture>

>       <!-- <source srcset="/web/updates/images/2020/11/cvd/0-regular@2x.avif 2x, /web/updates/images/2020/11/cvd/0-regular.avif 1x" type="image/avif" />

Reviewer: Can we uncomment these so we use the AVIF and WebP versions?

Reviewer: Note that you don't need to specify `1x` — it's the default. So you could do:

```
<source srcset="/web/updates/images/2020/11/cvd/0-regular.avif, /web/updates/images/2020/11/cvd/0-regular@2x.avif 2x" type="image/avif">
```

and the same for the other pics

Author: WebFu has a presubmit hook to disallow AVIF and WebP formats.

      <source srcset="/web/updates/images/2020/11/cvd/0-regular@2x.webp 2x, /web/updates/images/2020/11/cvd/0-regular.webp 1x" type="image/webp" /> -->

>       <img srcset="/web/updates/images/2020/11/cvd/0-regular@2x.png 2x, /web/updates/images/2020/11/cvd/0-regular.png 1x"
>         alt="A colorful picture of melted crayons, with no color vision deficiencies simulated"/>

Reviewer: This doesn't have a `src` attribute meaning it won't display in older browsers. Let's do it like this:

 ```suggestion
      <img src="/web/updates/images/2020/11/cvd/0-regular.png"
        srcset="/web/updates/images/2020/11/cvd/0-regular@2x.png 2x"
        alt="A colorful picture of melted crayons, with no color vision deficiencies simulated">
```

Also, we should probably add `width` and `height` to reduce CLS

    </picture>
  <figcaption>

>     <b>Figure 1</b>. Description: A colorful <a href="https://unsplash.com/photos/keMdIzsNzsM">picture of melted crayons</a>, with no color vision deficiencies simulated.

Reviewer: Sorry for the confusion — I left these `Figure X: Description:` notes in the doc just to clarify how I planned to use it. Since we're using `<figcaption>` it's already clear these are descriptions so let's drop this:

```suggestion
    A colorful <a href="https://unsplash.com/photos/keMdIzsNzsM">picture of melted crayons</a>, with no color vision deficiencies simulated.
```

  </figcaption>
</figure>

<figure>
  <picture>

>     <!-- <source srcset="/web/updates/images/2020/11/cvd/1-achromatopsia@2x.avif 2x, /web/updates/images/2020/11/cvd/1-achromatopsia.avif 1x" type="image/avif" />

Reviewer: (same comments as above, for all the other images)

    <source srcset="/web/updates/images/2020/11/cvd/1-achromatopsia@2x.webp 2x, /web/updates/images/2020/11/cvd/1-achromatopsia.webp 1x" type="image/webp" /> -->
    <img srcset="/web/updates/images/2020/11/cvd/1-achromatopsia@2x.png 2x, /web/updates/images/2020/11/cvd/1-achromatopsia.png 1x"
      alt="The impact of simulating achromatopsia on a colorful picture of melted crayons."/>
  </picture>
  <figcaption>

>     <b>Figure 2</b>. Description: The impact of simulating achromatopsia on a colorful picture of melted crayons.

Reviewer: ```suggestion
    The impact of simulating achromatopsia on a colorful picture of melted crayons.
```

  </figcaption>
</figure>

<figure>
  <picture>
    <!-- <source srcset="/web/updates/images/2020/11/cvd/2-deuteranopia@2x.avif 2x, /web/updates/images/2020/11/cvd/2-deuteranopia.avif 1x" type="image/avif" />
    <source srcset="/web/updates/images/2020/11/cvd/2-deuteranopia@2x.webp 2x, /web/updates/images/2020/11/cvd/2-deuteranopia.webp 1x" type="image/webp" /> -->
    <img srcset="/web/updates/images/2020/11/cvd/2-deuteranopia@2x.png 2x, /web/updates/images/2020/11/cvd/2-deuteranopia.png 1x"
      alt="The impact of simulating deuteranopia on a colorful picture of melted crayons."/>
  </picture>
  <figcaption>

>     <b>Figure 3</b>. Description: The impact of simulating deuteranopia on a colorful picture of melted crayons.

Reviewer: ```suggestion
    The impact of simulating deuteranopia on a colorful picture of melted crayons.
```

  </figcaption>
</figure>

<figure>
  <picture>
    <!-- <source srcset="/web/updates/images/2020/11/cvd/3-protanopia@2x.avif 2x, /web/updates/images/2020/11/cvd/3-protanopia.avif 1x" type="image/avif" />
    <source srcset="/web/updates/images/2020/11/cvd/3-protanopia@2x.webp 2x, /web/updates/images/2020/11/cvd/3-protanopia.webp 1x" type="image/webp" /> -->
    <img srcset="/web/updates/images/2020/11/cvd/3-protanopia@2x.png 2x, /web/updates/images/2020/11/cvd/3-protanopia.png 1x"
      alt="The impact of simulating protanopia on a colorful picture of melted crayons."/>
  </picture>
  <figcaption>

>     <b>Figure 4</b>. Description: The impact of simulating protanopia on a colorful picture of melted crayons.

Reviewer: ```suggestion
    The impact of simulating protanopia on a colorful picture of melted crayons.
```

  </figcaption>
</figure>

<figure>
  <picture>
    <!-- <source srcset="/web/updates/images/2020/11/cvd/4-tritanopia@2x.avif 2x, /web/updates/images/2020/11/cvd/4-tritanopia.avif 1x" type="image/avif" />
    <source srcset="/web/updates/images/2020/11/cvd/4-tritanopia@2x.webp 2x, /web/updates/images/2020/11/cvd/4-tritanopia.webp 1x" type="image/webp" /> -->
    <img srcset="/web/updates/images/2020/11/cvd/4-tritanopia@2x.png 2x, /web/updates/images/2020/11/cvd/4-tritanopia.png 1x"
      alt="The impact of simulating tritanopia on a colorful picture of melted crayons."/>
  </picture>
  <figcaption>

>     <b>Figure 5</b>. Description: The impact of simulating tritanopia on a colorful picture of melted crayons.

Reviewer: ```suggestion
    The impact of simulating tritanopia on a colorful picture of melted crayons.
```

  </figcaption>
</figure>

As a developer with regular vision, you might see DevTools display a bad contrast ratio for color pairs that visually look okay to you. This happens because the contrast ratio formulas take into account these color vision deficiencies! _You_ might still be able to read low-contrast text in some cases, but people with vision impairments don’t have that privilege.

By letting designers and developers simulate the effect of these vision deficiencies on their own web apps, we aim to provide the missing piece: not only can DevTools help you *find* and *fix* contrast issues, now you can also *understand* them!

> ## Simulating color vision deficiencies with HTML, CSS, SVG, and C++

Reviewer: ```suggestion

## Simulating color vision deficiencies with HTML, CSS, SVG, and C++

```

  <label><input type="radio" name="vision-deficiency" value="achromatopsia"> achromatopsia</label>
  <label><input type="radio" name="vision-deficiency" value="deuteranopia"> deuteranopia</label>
  <label><input type="radio" name="vision-deficiency" value="protanopia"> protanopia</label>
  <label><input type="radio" name="vision-deficiency" value="tritanopia"> tritanopia</label>
</div>
<script type="module">
  const wrapper = document.querySelector('div');
  const filtered = document.querySelector('.filtered');

  wrapper.addEventListener('click', (event) => {
    const target = event.target;
    if (target.matches('input')) {
      const value = target.value === 'none' ? 'none' : `url(#${target.value})`;
      // TODO: use `attributeStyleMap` once Safari and Firefox support it.
      filtered.style.filter = value;
    }
  });
</script>
{% endframebox %}

If we wanted to, we could build our DevTools feature as follows: when the user emulates a vision deficiency in the DevTools UI, we inject the SVG filter into the inspected document, and then we apply the filter style on the root element. However, there are several problems with that approach:

> The page might already have a filter on its root element, which our code might then override.

Reviewer: This should be a list

                           0.000  0.000  0.000  1.000  0.000">
  </filter>
</svg>
```

### Avoiding the in-document SVG dependency

Let’s start with part 2: how can we avoid adding the SVG to the DOM? One idea is to move it to a separate SVG file. We can copy the `<svg>…</svg>` from the above HTML and save it as `filter.svg` — but we need to make some changes first! Inline SVG in HTML follows the HTML parsing rules. That means you can get away with things like [omitting quotes around attribute values in some cases](https://mathiasbynens.be/notes/unquoted-attribute-values). However, SVG in separate files is supposed to be valid XML — and XML parsing is way more strict than HTML. Here’s our SVG-in-HTML snippet again:

```html
<svg>
  <filter id="deuteranopia">
    <feColorMatrix values="0.367  0.861 -0.228  0.000  0.000
                           0.280  0.673  0.047  0.000  0.000
                          -0.012  0.043  0.969  0.000  0.000
                           0.000  0.000  0.000  1.000  0.000">
  </filter>
</svg>
```

To make this valid standalone SVG (and thus XML), we need to make two changes. Can you guess which?

```xml

> <svg xmlns="http://www.w3.org/2000/svg">

Reviewer: I guess there's no way to highlight specific lines, or even characters, in WebFu code snippets?

Author: Possible, see the escape version of the code with `<pre>` tag and code highlight with `<strong>`.

    case VisionDeficiency::kTritanopia:
      return CreateFilterDataUrl("…");
    case VisionDeficiency::kNoVisionDeficiency:
      NOTREACHED();
      return "";
  }
}
```

Note that this technique gives us access to the full power of SVG filters without having to re-implement anything or re-invent any wheels. We’re implementing a Blink Renderer feature, but we’re doing so by leveraging the Web Platform.

Okay, so we’ve figured out how to construct SVG filters and turn them into data URLs that we can use within our CSS `filter` property value. Can you think of a problem with this technique? It turns out, we can’t actually _rely_ on the data URL being loaded in all cases, since the target page might have [a `Content-Security-Policy`](/web/fundamentals/security/csp) that blocks data URLs. Our final Blink-level implementation takes special care to bypass CSP for these “internal” data URLs during loading.

Edge cases aside, we’ve made some good progress. Because we no longer depend on inline `<svg>` being present in the same document, we’ve effectively reduced our solution to just a single self-contained CSS `filter` property definition. Great! Now let’s get rid of that too.

> 

Reviewer: ```suggestion
```

      InitialStyleForElement(GetDocument());
  viewport_style->SetZIndex(0);
  viewport_style->SetIsStackingContextWithoutContainment(true);
  viewport_style->SetDisplay(EDisplay::kBlock);
  viewport_style->SetPosition(EPosition::kAbsolute);
  viewport_style->SetOverflowX(EOverflow::kAuto);
  viewport_style->SetOverflowY(EOverflow::kAuto);
  // …
  return viewport_style;
}
```

You don't need to understand C++ or the intricacies of Blink’s Style engine to see that this code handles the viewport’s (or more accurately: the initial containing block’s) `z-index`, `display`, `position`, and `overflow`. Those are all concepts you might be familiar with from CSS! There’s some other magic related to stacking contexts, which doesn’t _directly_ translate to a CSS property, but overall you could think of this `viewport` object as something that can be styled using CSS from within Blink, just like a DOM element — except it’s not part of the DOM.

*This gives us exactly what we want!* We can apply our `filter` styles to the `viewport` object, which visually affects the rendering, without interfering with the observable page styles or the DOM in any way.

> 

Reviewer: ```suggestion
```

## Conclusion {: #conclusion }

To recap our little journey here, we started out by building a prototype using web technology instead of C++, and then started working on moving parts of it to the Blink Renderer.

> We first made our prototype more self-contained by inlining data URLs.

Reviewer: should be a list

{# wf_published_on: 2020-11-13 #}
{# wf_tags: devtools-blog #}
{# wf_featured_image: /web/updates/images/2020/11/cvd/cvd.png #}
{# wf_featured_snippet: Simulating color vision deficiencies in the Blink Renderer. #}
{# wf_blink_components: N/A #}

# Simulating color vision deficiencies in the Blink Renderer {: .page-title }

{% include "web/_shared/contributors/mathiasbynens.html" %}

This article describes why and how we implemented color vision deficiency simulation in DevTools and the Blink Renderer.

Note: If you prefer watching a presentation over reading articles, then enjoy the video below! If not, skip the video and read on.

<div class="video-wrapper-full-width">
  <iframe class="devsite-embedded-youtube-video" data-video-id="LnapsmokM_4" allowfullscreen></iframe>
</div>

## Background: bad color contrast {: #background }

[Low-contrast text](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) is the most common automatically-detectable accessibility issue on the web.

![A list of common accessibility issues on the web. Low-contrast text is by far the most common issue.](/web/updates/images/2020/11/cvd/common-a11y-issues.svg)

> According to [WebAIM’s accessibility analysis of the top 1-million web sites](https://webaim.org/projects/million/update#wcag:~:text=Low%20contrast%20text%0985.3%25%0986.1%25), over *86%* of home pages have low contrast. On average, each home page has [36 distinct instances](https://webaim.org/projects/million/#contrast:~:text=On%20average%2C%20home%20pages%20had%2036%20distinct%20instances%20of%20low%2Dcontrast%20text.) of low-contrast text.

Reviewer: ```suggestion
According to [WebAIM’s accessibility analysis of the top 1-million websites](https://webaim.org/projects/million/update#wcag:~:text=Low%20contrast%20text%0985.3%25%0986.1%25), over *86%* of home pages have low contrast. On average, each home page has [36 distinct instances](https://webaim.org/projects/million/#contrast:~:text=On%20average%2C%20home%20pages%20had%2036%20distinct%20instances%20of%20low%2Dcontrast%20text.) of low-contrast text.
```

## Simulating color vision deficiencies with HTML, CSS, SVG, and C++

Before we dive into the Blink Renderer implementation of our feature, it helps to understand how you’d implement equivalent functionality using web technology.

You can think of each of these color vision deficiency simulations as an overlay covering the entire page. The Web Platform has a way to do that: CSS filters! With the CSS `filter` property, you can use some predefined filter functions, such as `blur`, `contrast`, `grayscale`, `hue-rotate`, and many more. For even more control, the `filter` property also accepts a URL which can point to a custom SVG filter definition:

```html
<style>
  :root {
    filter: url(#deuteranopia);
  }
</style>
<svg>
  <filter id="deuteranopia">
    <feColorMatrix values="0.367  0.861 -0.228  0.000  0.000
                           0.280  0.673  0.047  0.000  0.000
                          -0.012  0.043  0.969  0.000  0.000
                           0.000  0.000  0.000  1.000  0.000">
  </filter>
</svg>
```

> The above example uses a custom filter definition based on a color matrix. Conceptually, every pixel’s `[Red, Green, Blue, Alpha]` color value is matrix-multiplied to create a new color `[R', G', B', A']`.

Reviewer: ```suggestion
The above example uses a custom filter definition based on a color matrix. Conceptually, every pixel’s `[Red, Green, Blue, Alpha]` color value is matrix-multiplied to create a new color `[R′, G′, B′, A′]`.
```

project_path: /web/_project.yaml
book_path: /web/updates/_book.yaml
description: Simulating color vision deficiencies in the Blink Renderer. The Chrome DevTools engineering blog - by the developers who build the DevTools.

{# wf_updated_on: 2020-11-13 #}
{# wf_published_on: 2020-11-13 #}
{# wf_tags: devtools-blog #}
{# wf_featured_image: /web/updates/images/2020/11/cvd/cvd.png #}
{# wf_featured_snippet: Simulating color vision deficiencies in the Blink Renderer. #}
{# wf_blink_components: N/A #}

# Simulating color vision deficiencies in the Blink Renderer {: .page-title }

{% include "web/_shared/contributors/mathiasbynens.html" %}

This article describes why and how we implemented color vision deficiency simulation in DevTools and the Blink Renderer.

> 

Reviewer: Can we add something like:

```suggestion
**Note:** If you prefer watching a presentation over reading articles, then enjoy the video below! If not, skip the video and read on.

<div class="video-wrapper-full-width">
  <iframe class="devsite-embedded-youtube-video" data-video-id="LnapsmokM_4"
          data-showinfo="0" allowfullscreen>
  </iframe>
</div>
```

## Using DevTools to find, understand, and fix contrast issues {: #fix }

Chrome DevTools can help developers and designers to improve contrast and to pick more accessible color schemes for web apps:

- The Inspect Mode tooltip that appears on top of the web page [shows the contrast ratio](/web/updates/2020/05/devtools#a11y) for text elements.
- The DevTools color picker [calls out bad contrast ratios](/web/updates/2018/01/devtools#contrast) for text elements, [shows the recommended contrast line](/web/updates/2019/01/devtools#AAA) to help manually select better colors, and can even [suggest accessible colors](/web/updates/2020/08/devtools#accessible-color).
- Both [the CSS Overview panel](/web/updates/2020/10/devtools#css-overview) and [the Lighthouse Accessibility audit report](https://web.dev/color-contrast/) lists low-contrast text elements as found on your page.

<div class="video-wrapper-full-width">
  <iframe class="devsite-embedded-youtube-video" data-video-id="Mje2wYgPYP0"

>           data-autohide="1" data-showinfo="0" frameborder="0" allowfullscreen>

Reviewer: ```suggestion
          data-showinfo="0" allowfullscreen>
```

`data-autohide` was deprecated and now defaults to `1` so I removed it

  </iframe>
</div>

We’ve recently added a new tool to this list, and it’s a bit different from the others. The above tools mainly focus on *surfacing contrast ratio information* and giving you options to *fix* it. We realized that DevTools was still missing a way for developers to get a deeper *understanding* of this problem space. To address this, we implemented [vision deficiency simulation](/web/updates/2020/03/devtools#vision-deficiencies) in the DevTools Rendering tab.

<div class="video-wrapper-full-width">
  <iframe class="devsite-embedded-youtube-video" data-video-id="mK_XmFb8E_w"

>           data-autohide="1" data-showinfo="0" frameborder="0" allowfullscreen>

Reviewer: ```suggestion
          data-showinfo="0" allowfullscreen>
```

`frameborder` is presentational, and it's already handled through CSS

```html
<svg>
  <filter id="deuteranopia">
    <feColorMatrix values="0.367  0.861 -0.228  0.000  0.000
                           0.280  0.673  0.047  0.000  0.000
                          -0.012  0.043  0.969  0.000  0.000
                           0.000  0.000  0.000  1.000  0.000">
  </filter>
</svg>
```

To make this valid standalone SVG (and thus XML), we need to make two changes. Can you guess which?

<!-- ```xml
<svg xmlns="http://www.w3.org/2000/svg">
  <filter id="deuteranopia">
    <feColorMatrix values="0.367  0.861 -0.228  0.000  0.000
                           0.280  0.673  0.047  0.000  0.000
                          -0.012  0.043  0.969  0.000  0.000
                           0.000  0.000  0.000  1.000  0.000" />
  </filter>
</svg>
``` -->

> <pre class="prettyprint">
> &lt;svg <strong>xmlns=&quot;http://www.w3.org/2000/svg&quot;</strong>&gt;
>   &lt;filter id=&quot;deuteranopia&quot;&gt;
>     &lt;feColorMatrix values=&quot;0.367  0.861 -0.228  0.000  0.000
>                           0.280  0.673  0.047  0.000  0.000
>                           -0.012  0.043  0.969  0.000  0.000
>                           0.000  0.000  0.000  1.000  0.000&quot; <strong>/&gt;</strong>
>   &lt;/filter&gt;
> &lt;/svg&gt;
> </pre>

Reviewer: On the preview, the WebFu styles have low-contrast text except for the highlighted parts, so let's try highlighting it as normal, and just adding `<mark>` where applicable:

```suggestion
<pre class="prettyprint"><code class="html"><span class="tag"><span class="tag">&lt;svg </span></span><mark><span class="atn"><span class="atn">xmlns</span></span><span class="pun"><span class="pun">=</span></span><span class="atv"><span class="atv">"http://www.w3.org/2000/svg"</span></span></mark><span class="tag"><span class="tag">&gt;</span></span><span class="pln"><span class="pln"><br>&nbsp; </span></span><span class="tag"><span class="tag">&lt;filter</span></span><span class="pln"><span class="pln"> </span></span><span class="atn"><span class="atn">id</span></span><span class="pun"><span class="pun">=</span></span><span class="atv"><span class="atv">"deuteranopia"</span></span><span class="tag"><span class="tag">&gt;</span></span><span class="pln"><span class="pln"><br>&nbsp; &nbsp; </span></span><span class="tag"><span class="tag">&lt;feColorMatrix</span></span><span class="pln"><span class="pln"> </span></span><span class="atn"><span class="atn">values</span></span><span class="pun"><span class="pun">=</span></span><span class="atv"><span class="atv">"0.367 &nbsp;0.861 -0.228 &nbsp;0.000 &nbsp;0.000<br>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;0.280 &nbsp;0.673 &nbsp;0.047 &nbsp;0.000 &nbsp;0.000<br>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -0.012 &nbsp;0.043 &nbsp;0.969 &nbsp;0.000 &nbsp;0.000<br>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;0.000 &nbsp;0.000 &nbsp;0.000 &nbsp;1.000 &nbsp;0.000"</span></span><mark><span class="tag"><span class="tag"> /&gt;</span></span></mark><span class="pln"><span class="pln"><br>&nbsp; </span></span><span class="tag"><span class="tag">&lt;/filter&gt;</span></span><span class="pln"><span class="pln"><br></span></span><span class="tag"><span class="tag">&lt;/svg&gt;</span></span><span class="pln"><span class="pln"><br></span></span></code></pre>
```

project_path: /web/_project.yaml
book_path: /web/updates/_book.yaml
description: Simulating color vision deficiencies in the Blink Renderer. The Chrome DevTools engineering blog - by the developers who build the DevTools.

{# wf_updated_on: 2020-11-13 #}
{# wf_published_on: 2020-11-13 #}
{# wf_tags: devtools-blog #}
{# wf_featured_image: /web/updates/images/2020/11/cvd/cvd.png #}
{# wf_featured_snippet: Simulating color vision deficiencies in the Blink Renderer. #}
{# wf_blink_components: N/A #}

# Simulating color vision deficiencies in the Blink Renderer {: .page-title }

{% include "web/_shared/contributors/mathiasbynens.html" %}

This article describes why and how we implemented color vision deficiency simulation in DevTools and the Blink Renderer.

> **Note:** If you prefer watching a presentation over reading articles, then enjoy the video below! If not, skip the video and read on.

Reviewer: Apparently I was overcomplicating things :D This is how to get the special note style to show up:

```suggestion

Note: If you prefer watching a presentation over reading articles, then enjoy the video below! If not, skip the video and read on.
```

<div class="video-wrapper-full-width">

>   <iframe class="devsite-embedded-youtube-video" data-video-id="LnapsmokM_4"
>           data-showinfo="0" allowfullscreen>
>   </iframe>

Reviewer: Ooh, apparently `showinfo` was deprecated too: https://developers.google.com/youtube/player_parameters#release_notes_08_23_2018

Most of the snippets on WebFu could probably be cleaned up but for now I'll just fix it here

```suggestion
  <iframe class="devsite-embedded-youtube-video" data-video-id="LnapsmokM_4" allowfullscreen></iframe>
```

</div>

> ## Background: bad color contrast {: #background }

Reviewer: ```suggestion

## Background: bad color contrast {: #background }
```

## Using DevTools to find, understand, and fix contrast issues {: #fix }

Chrome DevTools can help developers and designers to improve contrast and to pick more accessible color schemes for web apps:

- The Inspect Mode tooltip that appears on top of the web page [shows the contrast ratio](/web/updates/2020/05/devtools#a11y) for text elements.
- The DevTools color picker [calls out bad contrast ratios](/web/updates/2018/01/devtools#contrast) for text elements, [shows the recommended contrast line](/web/updates/2019/01/devtools#AAA) to help manually select better colors, and can even [suggest accessible colors](/web/updates/2020/08/devtools#accessible-color).
- Both [the CSS Overview panel](/web/updates/2020/10/devtools#css-overview) and [the Lighthouse Accessibility audit report](https://web.dev/color-contrast/) lists low-contrast text elements as found on your page.

<div class="video-wrapper-full-width">

>   <iframe class="devsite-embedded-youtube-video" data-video-id="Mje2wYgPYP0"
>           data-showinfo="0" allowfullscreen>
>   </iframe>

Reviewer: ```suggestion
  <iframe class="devsite-embedded-youtube-video" data-video-id="Mje2wYgPYP0" allowfullscreen></iframe>
```

</div>

We’ve recently added a new tool to this list, and it’s a bit different from the others. The above tools mainly focus on *surfacing contrast ratio information* and giving you options to *fix* it. We realized that DevTools was still missing a way for developers to get a deeper *understanding* of this problem space. To address this, we implemented [vision deficiency simulation](/web/updates/2020/03/devtools#vision-deficiencies) in the DevTools Rendering tab.

<div class="video-wrapper-full-width">

>   <iframe class="devsite-embedded-youtube-video" data-video-id="mK_XmFb8E_w"
>           data-showinfo="0" allowfullscreen>
>   </iframe>

Reviewer: ```suggestion
  <iframe class="devsite-embedded-youtube-video" data-video-id="mK_XmFb8E_w" allowfullscreen></iframe>
```
